### ISSUE TITLE
Offline deal imports not working on M1.3.5: Market node crash

### ISSUE THREAD

**Issue by benjaminh83** [filecoin-project/lotus#7144]
### Checklist

- [X] This is **not** a security-related bug/issue. If it is, please follow please follow the [security policy](https://github.com/filecoin-project/lotus/security/policy).
- [X] This is **not** a question or a support request. If you have any lotus related questions, please ask in the [lotus forum](https://github.com/filecoin-project/lotus/discussions).
- [X] This is **not** a new feature request. If it is, please file a [feature request](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Ffeature&template=feature_request.yml) instead.
- [X] This is **not** an enhancement request. If it is, please file a [improvement suggestion](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Fenhancement&template=enhancement.yml) instead.
- [X] I **have** searched on the [issue tracker](https://github.com/filecoin-project/lotus/issues) and the [lotus forum](https://github.com/filecoin-project/lotus/discussions), and there is no existing related issue or discussion.
- [X] I am running the [`Latest release`](https://github.com/filecoin-project/lotus/releases), or the most recent RC(release canadiate) for the upcoming release or the dev branch(master), or have an issue updating to any of these.
- [X] I did not make any code changes to lotus.

### Lotus component

- [ ] lotus daemon - chain sync
- [ ] lotus miner - mining and block production
- [ ] lotus miner/worker - sealing
- [ ] lotus miner - proving(WindowPoSt)
- [X] lotus miner/market - storage deal
- [ ] lotus miner/market - retrieval deal
- [ ] lotus miner/market - data transfer
- [ ] lotus client
- [ ] lotus JSON-RPC API
- [ ] lotus message management (mpool)
- [ ] Other

### Lotus Version

```text
$ lotus-miner version
Daemon:  1.11.1-m1.3.5+mainnet+git.3ff8e256b+api1.2.0
Local: lotus-miner version 1.11.1-m1.3.5+mainnet+git.3ff8e256b

$ lotus-miner --call-on-markets version
Daemon:  1.11.1-m1.3.5+mainnet+git.3ff8e256b+api1.2.0
Local: lotus-miner version 1.11.1-m1.3.5+mainnet+git.3ff8e256b
```

### Describe the Bug

I know others have been talking about issues related to market split and offline deal imports. I just wonder - did you actually test offline deal imports on the systems, when you tested the code. I'm still on M1.3.5 and running split on separate machines. I do the offline deal import on the market node machine, where the downloaded file is also located. Crashes market node with 100% certainty:
```
$ lotus-miner --call-on-markets storage-deals import-data bafyreihgxl4gqmcqqizlxdcejnowneyk4vuu6e2kqn4mtnvyywa7jagy7q /mnt/md4/stage/IRIS_313-11198_IRIS_313-11199_IRIS_313-11200_IRIS_313-11201.car 
2021-08-20T10:56:39.365+0200	WARN	rpc	go-jsonrpc@v0.1.4-0.20210217175800-45ea43ac2bec/websocket.go:678	failed to write close message: write tcp 192.168.20.42:43500->192.168.20.42:8787: write: broken pipe
ERROR: handler: websocket connection closed
```
I have attached the market logs from a fresh run on the market node. Just to keep things clean. No GO dump is provided on the crash. I can confirm that online deals (that actually manage to download) is going fine through the AP and sealing - been doing a lot on this release so far.

I'm not a fan of setting up NFS and other workarounds - I would much more like the issues to be resolved in lotus, as setups becomes very complex with a lot of NFS shares crossing between the main miner and market node. I think this should be able to flow nicely, and hopefully directly from market node to AP worker soon!

### Logging Information

```text
will be added
```

### Repo Steps

Basically just try to run an offline import on the market node: 
`lotus-miner storage-deals import-data xxx`

**Comment by benjaminh83** [comment]
[market-m1.3.5-log5.log](https://github.com/filecoin-project/lotus/files/7020721/market-m1.3.5-log5.log)

**Comment by aarshkshah1992** [comment]
cc @jacobheun - this should be high priority until we decide otherwise based on our invstigation.

**Comment by benjaminh83** [comment]
Adding in logs from main miner in the period around the corresponding crash on the market node. I do not see anything on the log that suggest it has issues with the offline deal import happening on the market node: 
```
$ cat /mnt/seal/logs/miner-m1.3.5-log1.log | grep 2021-08-20T10:5
2021-08-20T10:50:01.832+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:50:06.101+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 30, "forRound": 1038101, "baseEpoch": 1038100, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133945, "lookbackEpochs": 900, "networkPowerAtLookback": "10787997185551925248", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:50:36.516+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 8, "forRound": 1038102, "baseEpoch": 1038101, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133946, "lookbackEpochs": 900, "networkPowerAtLookback": "10787997185551925248", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:51:01.759+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:51:06.139+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 99, "forRound": 1038103, "baseEpoch": 1038102, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133947, "lookbackEpochs": 900, "networkPowerAtLookback": "10788057693051191296", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:51:36.038+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 9, "forRound": 1038104, "baseEpoch": 1038103, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133948, "lookbackEpochs": 900, "networkPowerAtLookback": "10788176027990130688", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:52:02.213+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:52:06.162+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 56, "forRound": 1038105, "baseEpoch": 1038104, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133949, "lookbackEpochs": 900, "networkPowerAtLookback": "10788179979360043008", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:52:36.450+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 13, "forRound": 1038106, "baseEpoch": 1038105, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133950, "lookbackEpochs": 900, "networkPowerAtLookback": "10788180048079519744", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:53:01.619+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:53:06.106+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 34, "forRound": 1038107, "baseEpoch": 1038106, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133951, "lookbackEpochs": 900, "networkPowerAtLookback": "10788180048079519744", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:53:36.160+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 11, "forRound": 1038108, "baseEpoch": 1038107, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133952, "lookbackEpochs": 900, "networkPowerAtLookback": "10788179910640566272", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:54:01.717+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:54:06.387+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 33, "forRound": 1038109, "baseEpoch": 1038108, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133953, "lookbackEpochs": 900, "networkPowerAtLookback": "10788258731880382464", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:54:36.266+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 5, "forRound": 1038110, "baseEpoch": 1038109, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133954, "lookbackEpochs": 900, "networkPowerAtLookback": "10788339614704500736", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:55:01.506+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:55:06.079+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 48, "forRound": 1038111, "baseEpoch": 1038110, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133955, "lookbackEpochs": 900, "networkPowerAtLookback": "10788339683423977472", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:55:36.035+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 6, "forRound": 1038112, "baseEpoch": 1038111, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133956, "lookbackEpochs": 900, "networkPowerAtLookback": "10788325699010461696", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:56:01.790+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:56:06.251+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 55, "forRound": 1038113, "baseEpoch": 1038112, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133957, "lookbackEpochs": 900, "networkPowerAtLookback": "10788391910226296832", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:56:36.038+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 9, "forRound": 1038114, "baseEpoch": 1038113, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133958, "lookbackEpochs": 900, "networkPowerAtLookback": "10788496638708842496", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:57:02.202+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:57:06.501+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 69, "forRound": 1038115, "baseEpoch": 1038114, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133959, "lookbackEpochs": 900, "networkPowerAtLookback": "10788496741788057600", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:57:36.368+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 7, "forRound": 1038116, "baseEpoch": 1038115, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133960, "lookbackEpochs": 900, "networkPowerAtLookback": "10788499215689220096", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:58:02.142+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:58:06.240+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 43, "forRound": 1038117, "baseEpoch": 1038116, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133961, "lookbackEpochs": 900, "networkPowerAtLookback": "10788499215689220096", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:58:36.042+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 5, "forRound": 1038118, "baseEpoch": 1038117, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133962, "lookbackEpochs": 900, "networkPowerAtLookback": "10788579926714646528", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:59:02.303+0200    DEBUG   advmgr  sector-storage/sched.go:354     SCHED 0 queued; 6 open windows
2021-08-20T10:59:06.376+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 50, "forRound": 1038119, "baseEpoch": 1038118, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133963, "lookbackEpochs": 900, "networkPowerAtLookback": "10788579926714646528", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
2021-08-20T10:59:36.049+0200    INFO    miner   miner/miner.go:478      completed mineOne       {"tookMilliseconds": 7, "forRound": 1038120, "baseEpoch": 1038119, "baseDeltaSeconds": 6, "nullRounds": 0, "lateStart": false, "beaconEpoch": 1133964, "lookbackEpochs": 900, "networkPowerAtLookback": "10788579926714646528", "minerPowerAtLookback": "360741606686720", "isEligible": true, "isWinner": false, "error": null}
```

**Comment by raulk** [comment]
This was somehow related to the `TMPDIR` environment var. This is a [standard POSIX var](https://en.wikipedia.org/wiki/TMPDIR) that Golang uses create temporary files. I'm not sure where in m1.3.5 we were demanding a temp file from the filesystem.

We do use the `FileStore#CreateTemp()` method to allocate a temp file in the offline import flow, but that creates a temp file in a root directory that's specified by Lotus when constructing the `FileStore`. **It does not ask the OS for a temp path (so not sure where TMPDIR was being used).** In m1.3.5, this tmp dir base was `$LOTUS_MARKETS_PATH/tmp`:

![image](https://user-images.githubusercontent.com/1101242/130281912-79dc4e57-ea86-4e5d-855c-0331bf85316d.png)

NOTE: In master, this has been changed to `$LOTUS_MARKETS_HOME/deal-staging`.

---

@benjaminh83 reports that he experienced this problem with `TMPDIR` being empty, likely because `/tmp` ([default OS tmp dir on Linux](https://github.com/golang/go/blob/master/src/os/file_unix.go#L329-L339) when TMPDIR is not set) didn't have sufficient space for his import.

He also says that the command crashed after a while, indicating that it may have been working on that path until the disk got full.

When he changes `TMPDIR` to an alternate location with 2TiB space, the command works well.

---

I'm closing this since the issue is fixed, but what still isn't clear to me is where `TMPDIR` is used.
